redhat-developer/rhdh-operator
# Remove usage of `kube-rbac-proxy` and switch to the built-in `WithAuthenticationAndAuthorization` filter instead

## Description

This removes all usage of the `kube-rbac-proxy` sidecar container, which was used to protect the `/metrics` endpoint exposed by the controller. This ensured that only authorized users and service accounts could access potentially sensitive metrics data.
Instead, it is now recommended to leverage the built-in [`WithAuthenticationAndAuthorization`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/metrics/filters#WithAuthenticationAndAuthorization) filter which provides a similar authn/authz protection of this endpoint but without the need for an extra container.
More context in https://github.com/kubernetes-sigs/controller-runtime/issues/2073 and https://book.kubebuilder.io/reference/metrics#how-the-metrics-endpoint-can-be-protected-

## Which issue(s) does this PR fix or relate to

- Fixes https://issues.redhat.com/browse/RHIDP-4236
- Obsoletes https://issues.redhat.com/browse/RHIDP-2830
- Obsoletes https://issues.redhat.com/browse/RHIDP-2298

## PR acceptance criteria


## How to test changes / Special notes to the reviewer

From this PR:

- The Operator Deployment should now have a single container if you deploy it with or without OLM
- You can run the E2E test case added in https://github.com/redhat-developer/rhdh-operator/pull/338/files#diff-7d1fc1e232b55677582b51e2623aae6d8c15b5f0960ac6100fc7d733cb795dac and make sure it passes:
```sh
# Download Ginkgo locally if needed
$ make ginkgo
$ IMG=quay.io/rhdh-community/operator:0.4.0-pr-338 \
  ./bin/ginkgo -v -p -focus 'operator metrics endpoint' tests/e2e

[...]
Ran 3 of 9 Specs in 48.442 seconds
SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 6 Skipped

Ginkgo ran 1 suite in 50.800777439s
Test Suite Passed
```
